core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
+  /** Indicates if we are in the middle of processing some event */
+  private val processingEvent = new AtomicBoolean(false)
+  /**
+   * Indicates if the event executor is started. The executor thread will be
+   * blocked on the condition variable until the event executor is started to
+   * guarantee that we do not process any event before starting the event executor.
+   */
+  private val isStarted = new AtomicBoolean(false)
+  private val lock = new ReentrantLock()
+  /** Condition variable which is signaled once the event executor is started */
+  private val startCondition: Condition = lock.newCondition
+
+  def addListener(listener: L): Unit = {
+    listeners.add(listener)
+  }
+
+  def removeListener(listener: L): Unit = {
+    listeners.remove(listener)
+  }
+
+  def start(): Unit = {
+    isStarted.set(true)
+    lock.lock()
+    try {
+      startCondition.signalAll()
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  private[this] def waitForStart(): Unit = {
+    lock.lock()
+    try {
+      while (!isStarted.get()) {
+          startCondition.await()
+        }
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  def submit(task: Runnable): Unit = {
+    try {
+      executorService.submit(new Runnable {
+        override def run(): Unit = LiveListenerBus.withinListenerThread.withValue(true) {
+          waitForStart()
+          processingEvent.set(true)
+          task.run()
+          processingEvent.set(false)
+      }
+      })
+    } catch {
+      case e: RejectedExecutionException =>
+        droppedEventsCounter.incrementAndGet()
+        if (System.currentTimeMillis() - lastReportTimestamp >= 60 * 1000) {
+          val droppedEvents = droppedEventsCounter.get
+          // There may be multiple threads trying to decrease droppedEventsCounter.
+          // Use "compareAndSet" to make sure only one thread can win.
+          // And if another thread is increasing droppedEventsCounter, "compareAndSet" will fail and
+          // then that thread will update it.
+          if (droppedEventsCounter.compareAndSet(droppedEvents, 0)) {
+            droppedEventsCounter.set(0)
+            val prevLastReportTimestamp = lastReportTimestamp
+            lastReportTimestamp = System.currentTimeMillis()
+            logError(s"Dropping $droppedEvents SparkListenerEvent since " +
+              new java.util.Date(prevLastReportTimestamp) +
+              " because no remaining room in event queue. This likely means" +
+              s" $listenerName event processor is too slow and cannot keep up " +
+              "with the rate at which tasks are being started by the scheduler.")
+          }
+        }
+    }
+  }
+
+  private[this] def isProcessingEvent: Boolean = processingEvent.get()
+
+  def isEmpty: Boolean = {
+    executorService.getQueue.size() == 0 && !isProcessingEvent
+  }
+
+  def stop(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10, TimeUnit.SECONDS)
```

Comment:
This changes the current behavior slightly. Right now the bus will wait until the listener thread stops:

    listenerThread.join()

Here you might return before things are really shut down. It's unlikely to happen, but if it does happen, it can lead to weird issues. Might be better to try `shutdownNow()` if this call fails, which at least makes a best effort to interrupt existing tasks that might be slow.